File: geom.erl
-module(geom).
-export([area/1, area/3]).

area({Shape, A, B}) ->
    area(Shape, A, B).

area(rectangle, Length, Width)  when Length >= 0, Width >= 0  -> Length * Width;
area(ellipse, A, B)             when A >= 0, B >= 0           -> math:pi() * A * B;
area(triangle, Base, Height)    when Base >= 0, Height >= 0   -> Base * Height / 2;
area(_, _, _) -> 0.

File: Geom.ex
defmodule Geom do
  @vsn 1.1

  @spec area(atom(), number(), number()) :: number()
  def area(shape, a, b) when a >= 0 and b >= 0 do
    case shape do
      :rectangle -> a * b
      :triangle  -> a * b /2
      :ellipse   -> :math.pi * a * b
      _          -> 0
    end
  end

  @spec area( { atom(), number(), number() } ) :: number()
  def area({shape, a, b }) do
    area(shape, a, b)
  end
end
